// @storybook/addon-vitest/dist/_node-chunks/vitest.config.4.template-BHJGNR5T.js
import CJS_COMPAT_NODE_URL_iir1e5cw7r from 'node:url';
import CJS_COMPAT_NODE_PATH_iir1e5cw7r from 'node:path';
import CJS_COMPAT_NODE_MODULE_iir1e5cw7r from "node:module";

var __filename = CJS_COMPAT_NODE_URL_iir1e5cw7r.fileURLToPath(import.meta.url);
var __dirname = CJS_COMPAT_NODE_PATH_iir1e5cw7r.dirname(__filename);
var require = CJS_COMPAT_NODE_MODULE_iir1e5cw7r.createRequire(import.meta.url);

// ------------------------------------------------------------
// end of CJS compatibility banner, injected by Storybook's esbuild configuration
// ------------------------------------------------------------
import "./chunk-JXLUGFAL.js";

// raw:../templates/vitest.config.4.template.ts
var vitest_config_4_template_default = `import path from 'node:path';
import { fileURLToPath } from 'node:url';

import { defineConfig } from 'vitest/config';

import { storybookTest } from '@storybook/addon-vitest/vitest-plugin';

import { playwright } from '@vitest/browser-playwright';

const dirname =
  typeof __dirname !== 'undefined' ? __dirname : path.dirname(fileURLToPath(import.meta.url));

// More info at: https://storybook.js.org/docs/next/writing-tests/integrations/vitest-addon
export default defineConfig({
  test: {
    projects: [
      {
        extends: true,
        plugins: [
          // The plugin will run tests for the stories defined in your Storybook config
          // See options at: https://storybook.js.org/docs/next/writing-tests/integrations/vitest-addon#storybooktest
          storybookTest({ configDir: path.join(dirname, 'CONFIG_DIR') }),
        ],
        test: {
          name: 'storybook',
          browser: {
            enabled: true,
            headless: true,
            provider: playwright({}),
            instances: [{ browser: 'chromium' }],
          },
        },
      },
    ],
  },
});
`;
export {
  vitest_config_4_template_default as default
};
